src/backend/lib/upgrade-server.js

```diff
@@ -0,0 +1,271 @@
+    this.storage = storage;
+    this.stateLock = new RWLock();
+    this.state = null;
+    this.context = null;
+    this.setState(UpgradeState.Server.Idle, null);
+    this.port = port;
+    this.server = http.createServer((req, res) => {
+      log("got http request", req.url, req.headers.host);
+      if (!this.handleHttpRequest(req, res)) {
+        log("invalid route", req.url);
+        res.statusCode = 404;
+        res.end();
+      }
+    });
+    this.announceInterval = null;
+
+    // In-progress uploads
+    this.uploads = [];
+  }
+
+  // XXX: public; also no concurrency guards!
+  setState(to, context) {
+    this.state = to;
+    this.context = context;
+    this.emit("state", this.state, this.context);
+  }
+
+  // Callback<Void> -> Void
+  share(cb) {
+    this.stateLock.writeLock(release => {
+      const self = this;
+      function done() {
+        release();
+        if (cb) cb();
+      }
+
+      if (this.state !== UpgradeState.Server.Idle) {
+        log("tried to share() when not in idle state");
+        return done();
+      }
+
+      log("starting discovery..");
+      const opts = {
+        server: [],
+        loopback: false,
+      };
+      this.discovery = discovery(opts);
+      this.server.listen(this.port, "0.0.0.0", () => {
+        log("announcing on", DISCOVERY_KEY);
+        this.announceInterval = setInterval(() => {
+          this.discovery.announce(DISCOVERY_KEY, this.port);
+        }, 2000);
+        this.setState(UpgradeState.Server.Sharing, this.uploads);
+        done();
+      });
+    });
+  }
+
+  // Callback<Void> -> Void
+  drain(cb) {
+    this.stateLock.writeLock(release => {
+      const self = this;
+      function done(err) {
+        if (err) self.setState(UpgradeState.Server.Error, err);
+        release();
+        if (cb) cb(err);
+      }
+
+      if (this.state !== UpgradeState.Server.Sharing) {
+        log("tried to drain() when not in share state");
+        return done();
+      }
+
+      this.setState(UpgradeState.Server.Draining, this.uploads);
+      if (this.uploads.length === 0) {
+        log("drain: no current uploads; stopping immediately");
+        return stop.bind(this)();
+      }
+
+      // wait for all uploads to finish
+      log(`drain: waiting for ${this.uploads.length} uploads to finish`);
+      release();
+      this.on("upload-complete", oncomplete.bind(this));
+      function oncomplete() {
+        if (this.uploads.length === 0) {
+          log("drain: no uploads left; stopping now");
+          this.removeListener("upload", oncomplete);
+          stop.call(this);
+        } else {
+          log(
+            `drain: still waiting for ${this.uploads.length} uploads to finish`
+          );
+        }
+      }
+
+      function stop() {
+        log("drain: stopping..");
+
+        clearInterval(this.announceInterval);
+        this.announceInterval = null;
+
+        // XXX: 'unannounce' seems to always return an error here; ignoring.
```

Comment:
Yes, I checked the dns-discovery code, and there is a bug where this will always return an error (because `success` is never set to true on the unannounce path when no servers are defined) and will never emit `unannounce`. https://github.com/mafintosh/dns-discovery/blob/master/index.js#L425-L482